### PR TITLE
Refactor ZeebeAssertionsExtension

### LINKED ISSUE #62: Add error handling to the ZeebeAssertionsExtension
**Description**

In the `ZeebeAssertionsExtensions` we use reflection to find specific fields in tests. At the moment we assume these fields are always available. If they are not the `Optional.get()` function will throw default exceptions.

We should handle these default exceptions and provide detailed feedback about what went wrong (eg. what fields we are missing in the test class).

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

- Split code into separate methods
- Add error handling
- Add unit tests

## Related issues

closes #62

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually

Documentation:
* [ ] Javadoc has been written
* [ ] The documentation is updated

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
@korthout Would you be willing to review this PR since Peter is on holiday? I'm happy to answer any questions you might have.

**Comment by korthout** [comment]
Hi @remcowesterhoud. 👋  

Sure I'd love to help, but I'm having some trouble understanding these changes.

Currently there is one big commit that is called `refactor: Refactor ...`, but it doesn't really help me understand things. I have the following questions that I want to find answers to:
- Why do these changes resolve #62? Everything seems to indicate that this is a refactoring, and does not improve the error output.
- There are tests added, do these verify that the behavior is still the same after the refactoring?
- Did this test also pass before the refactorings?

To make my life easier, can you please split up the changes into smaller commits?

**Comment by remcowesterhoud** [comment]
Thanks @korthout. 

I've split the commits into 2 separate commits. One with the added error handling, and one with the refactor. Hope this helps!

To answer your questions:

**Why do these changes resolve Add error handling to the ZeebeAssertionsExtension #62? Everything seems to indicate that this is a refactoring, and does not improve the error output.**
Before we would use `get()` the `recordStreamOptional` and the `zeebeClientOptional`. We did not check whether or not these fields were present. With this change we will make sure they are present, or else throw an exception explaining that these fields are required.

**There are tests added, do these verify that the behavior is still the same after the refactoring?**
The error handling is new as explained in in the question above. Therefore, the unit tests are also new.

**Did this test also pass before the refactorings?**
See above answers 😄

**Comment by korthout** [comment]
@remcowesterhoud Thanks this helps a lot! 👍

File: src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Test
+    public void testExtensionThrowsExceptionWhenMissingRecordStreamSource() {
+      // given
+      final ZeebeAssertionsExtension extension = new ZeebeAssertionsExtension();
+      final ExtensionContext extensionContext = mock(ExtensionContext.class);
+      Mockito.when(extensionContext.getTestClass()).thenReturn(Optional.of(this.getClass()));
+      Mockito.when(extensionContext.getTestInstance()).thenReturn(Optional.of(this));
+
+      // when & then
```
Comment (korthout) [inline src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java:28]: 🙃  this `//when` comment belongs above the `Mockito.when` calls

File: src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.mockito.Mockito;
+
+class ZeebeAssertionsExtensionTest {
+
+  @Nested
+  class MissingRecordStreamSource {
+
+    private ZeebeClient client;
```
Comment (korthout) [inline src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java:18]: 🔧 Maybe add a comment to this field, explaining why it is necessary for this test case, i.e. in order for `RecordStreamSource` to be the only one that is missing it needs to have all other required fields, incl `ZeebeClient`.

File: src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+import io.camunda.zeebe.client.ZeebeClient;
+import java.util.Optional;
+import org.camunda.community.eze.RecordStreamSource;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.mockito.Mockito;
+
+class ZeebeAssertionsExtensionTest {
```
Comment (korthout) [inline src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java:13]: 👍  I really like the simplicity of these tests, they are easy to understand. 🏅

File: src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java [hunk 4]
```diff
@@ -11,50 +11,66 @@
...
+    ReflectionUtils.makeAccessible(recordStreamField);
+    final RecordStreamSource recordStreamSource =
+        (RecordStreamSource) recordStreamField.get(testInstance);
+    RecordStreamSourceStore.init(recordStreamSource);
+  }
+
+  private Field getRecordStreamField(final ExtensionContext extensionContext) {
     final Optional<Field> recordStreamOptional =
-        Arrays.stream(extensionContext.getTestClass().get().getDeclaredFields())
+        Arrays.stream(extensionContext.getRequiredTestClass().getDeclaredFields())
```
Comment (korthout) [inline src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java:46]: 🙃  This change it not a refactoring, i.e. it changes the error message if the test class is missing for the extensionContext. TBH, I'm not sure whether that can actually happen, which brings me to something actually interesting.

🔧  There is something to say for dealing with the case that the class or the instance is missing (`getTestClass().orElseThrow(...)`), in which case we could help the user with a helpful error message similar to the missing required fields. All this depends on whether the class or the instance (below) can ever be missing. If they can never be missing than I prefer `getRequiredTestClass()`.

File: src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java [hunk 5]
```diff
@@ -11,50 +11,66 @@
...
 import org.junit.platform.commons.util.ReflectionUtils;
 
-// TODO rewrite this mess
 public class ZeebeAssertionsExtension implements BeforeEachCallback, AfterEachCallback {
 
   private static final String KEY_ZEEBE_CLIENT = "ZEEBE_CLIENT";
 
   @Override
   public void beforeEach(final ExtensionContext extensionContext) throws Exception {
+    final Object testInstance = extensionContext.getRequiredTestInstance();
```
Comment (korthout) [inline src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java:20]: 🙃  For understandability, I'd move this line into both of the store methods, duplicated just like the `getRequiredTestClass()` calls. That way the beforeEach becomes more descriptive, first do this, then do that.

**Review by korthout — CHANGES_REQUESTED** [review]
👍  Thanks for splitting up the changes and your explanations, it made reviewing much more easy. 

I found some things important to raise and the rest are suggestions that are at a take-it-or-leave-it level. Please have a look.

Additionally, please update the pull request summary to reflect the actual changes.

File: src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  @Nested
+  class MissingRecordStreamSource {
+
+    private ZeebeClient client;
+
+    @Test
+    public void testExtensionThrowsExceptionWhenMissingRecordStreamSource() {
+      // given
+      final ZeebeAssertionsExtension extension = new ZeebeAssertionsExtension();
+      final ExtensionContext extensionContext = mock(ExtensionContext.class);
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply src/test/java/io/camunda/testing/extensions/ZeebeAssertionsExtensionTest.java:24]: Ah I should've specified that this project uses a Java 8 language level. That way users with that are still running on older java version are able to use it as well 😄

File: src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java [hunk 7]
```diff
@@ -22,7 +22,11 @@ public void beforeEach(final ExtensionContext extensionContext) throws Exception
         Arrays.stream(extensionContext.getTestClass().get().getDeclaredFields())
             .filter(field -> field.getType() == RecordStreamSource.class)
             .findFirst();
-    final Field recordStreamField = recordStreamOptional.get();
+    final Field recordStreamField = recordStreamOptional.orElseThrow(
+        () ->
+            new IllegalStateException(
+                "No RecordStreamSource has been found. Please make sure a "
+                    + "RecordStreamSource field has been declared in the test class."));
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java:53]: I was not aware about these guidelines. I have applied them to the error messages in this PR.

As for the other messages in this project, they are all assertions. We have decided on a way to formulate the assertion messages so we should keep these the way they are for now. If user feedback suggests that these message are unclear and need to be improved we can do this at a later stage.

File: src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java (hunk 4 above)
Comment (remcowesterhoud) [reply src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java:46]: I thought about this. I don't know in which scenario the test class could not be present. Maybe when using the extension on a non-testclass. But when that happens the `beforeEach()` wouldn't be executed in the first place.

I will stick with `getRequiredTestClass()` for now. If we ever run into this problem we can find a solution at that time.

**Comment by remcowesterhoud** [comment]
Thanks for your review @korthout! 🍪

I've commented / applied your suggestion.

File: src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java [hunk 8]
```diff
@@ -49,8 +49,8 @@ private Field getRecordStreamField(final ExtensionContext extensionContext) {
     return recordStreamOptional.orElseThrow(
         () ->
             new IllegalStateException(
-                "No RecordStreamSource has been found. Please make sure a "
-                    + "RecordStreamSource field has been declared in the test class."));
+                "Expected a field of type RecordStreamSource to be declared in the test class, "
+                    + "but none has been found"));
```
Comment (korthout) [inline src/main/java/io/camunda/testing/extensions/ZeebeAssertionsExtension.java:51]: 🔧 To be honest, now I miss the user hint. Would be nice to add it behind it.

Something like:
```
"Expected a field of type RecordStreamSource to be declared in the test class, "
    + "but none has been found. Please make sure a field of type RecordStreamSource "
    + "has been declared in the test class."
```

**Review by korthout — APPROVED** [review]
LGTM 🥨 

Please consider my last comment before merging

**Pull request merged**